src/support/Type.ts

```diff
@@ -4,13 +4,27 @@ import { StringHelper } from './StringHelper';
 import { TypeMetadata } from './metadata/TypeMetadata';
 import { Entity } from '../Entity';
 
-export function Type(type?: Function, jsonKey?: string) {
+export type Constructor<T> = { new(...args: any): T };
+
+// Types that can be passed as first argument to `EntityBuilder`.
+export type Buildable = Constructor<Entity> | typeof Object | typeof String | typeof Number | typeof Boolean;
+export type DefaultExportedBuildable = { default: Buildable };
+export type PackedBuildable = Buildable | DefaultExportedBuildable;
+
+export type BuildableResolverSync = () => PackedBuildable;
+export type BuildableResolverAsync = () => Promise<PackedBuildable>;
+export type BuildableResolver = BuildableResolverAsync | BuildableResolverSync;
+
+// Types that can be passed to @Type decorator factory.
+export type Typeable = Buildable | BuildableResolver;
+
+export function Type<T extends Typeable>(type: T, jsonKey?: string) {
     return function (target: Entity, key: string) {
         jsonKey = jsonKey ? jsonKey : (
             EntityBuilder.enableCamelConversion ? StringHelper.toSnake(key) : key
         );
 
-        const metadata = new TypeMetadata(target.constructor, key, jsonKey, type);
+        const metadata = new TypeMetadata(target.constructor as typeof Entity, key, jsonKey, type);
```

Comment:
I've seen use of both `Constructor<Entity>`/`Constructor<T>` and `typeof` – is it intentional to use both? How do they differ?